### PR TITLE
Add spec for flow framework refactoring

### PR DESCRIPTION
**Pull request by anmolbabu**
Add spec for flow framework refactoring

Signed-off-by: anmolbabu <anmolbudugutta@gmail.com>

### CODE REVIEW

**Comment by anmolbabu** [comment]
@shtripat @nthomas-redhat @r0h4n @brainfunked Please review

**Review by r0h4n — CHANGES_REQUESTED** [review]
spec name "monitoring_defs" is too vague.

File: specs/flow_defs_reorganization.adoc [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+
+== Use Cases
+
+* Monitoring code cannot be part of any component of tendrl core stack.
+  The monitoring stack needs to provide a flow for configuring the monitoring
+  agents maintained by tendrl.
+
+== Proposed change
+
+* The node-agent pushes a master definition /tendrl_definitions to etcd which
```
Comment (brainfunked) [inline specs/flow_defs_reorganization.adoc]: We should have a namespace `tendrl` in etcd, under which no directory or file needs to have the `tendrl` prefix.

File: specs/flow_defs_reorganization.adoc [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+  node_agent reads the master definitions in /tendrl_definitions in etcd
+  and subsequently reads all definitions in /tendrl_integrations.d/ on
+  seeing the line "include /tendrl_integrations.d/*".
+
+
+=== Alternatives
+
+* Append monitoring related definitions yaml to node_agents definitions
+  directory in etcd.
+
```
Comment (brainfunked) [inline specs/flow_defs_reorganization.adoc]: Please fix the formatting here. Don't think blank lines are allowed between list items.

**Review by brainfunked — CHANGES_REQUESTED** [review]
This specification also requires three other specifications:
* To include additional definition files via an `include` statement, that should be valid only in the master definition file for now.
* Concept of reserved namespaces and namespace overrides. Definition files loaded via include statements should be able to override existing namespaces and objects. However, some namespaces, for instance the global tendrl namespace, should be reserved and not be allowed to be overridden. These are the reserved namespaces.
* Compilation of the definition files together should be done via a definition refresh flow in the node agent. The compiled, merged namespaced representation of all the definition files combined is what every component should refer to.

File: specs/flow_defs_reorganization.adoc (hunk 2 above)
Comment (anmolbabu) [reply specs/flow_defs_reorganization.adoc]: Done

File: specs/flow_defs_reorganization.adoc (hunk 1 above)
Comment (anmolbabu) [reply specs/flow_defs_reorganization.adoc]: Done

File: specs/monitoring_defs.adoc [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  to create the definitions path instead of hard-coding it.
+* And from there the already existing flow framework as usual will extract and
+  execute the pre, post and the normal flow atoms.
+
+=== Alternatives
+
+* Append monitoring related definitions yaml to node_agents definitions
+  directory in etcd.
+
+    ** Drawbacks
```
Comment (shtripat) [inline specs/monitoring_defs.adoc]: I dont think you need tabs if you are using ```**``` and ```***```. They automatically would happen in final doc

**Review by shtripat — COMMENTED** [review]
I dont see any major comments as such. Looks fine to me as long as other comments are taken care.

**Comment by anmolbabu** [comment]
@brainfunked 
'To include additional definition files via an include statement, that should be valid only in the master definition file for now.
Concept of reserved namespaces and namespace overrides. Definition files loaded via include statements should be able to override existing namespaces and objects. However, some namespaces, for instance the global tendrl namespace, should be reserved and not be allowed to be overridden. These are the reserved namespaces.'
These are to be done as part of flow validation right?
@r0h4n Could you please help raising specs for these
'Compilation of the definition files together should be done via a definition refresh flow in the node agent. The compiled, merged namespaced representation of all the definition files combined is what every component should refer to.'
@r0h4n is this handled as pert of any of your PRs already? Probably as part of spec for:
'Every component will build an in-memory map of the definitions based on the namespace versions in etcd. A namepsace can be assembled using multiple yaml files. Every part of the namespace (objects, flows etc.) would be referenced based on the latest version available for it.'

**Comment by r0h4n** [comment]
Ill work on this one @anmolbabu , you can focus on the monitoring flows

**Comment by anmolbabu** [comment]
Closing this as @r0h4n has addressed it.

**Pull request closed without merging**